File: leetcode/539_Minimum_Time_Difference/539_Minimum_Time_Difference.java
class Solution {
    private int transform(String timePoint) {
        String[] times = timePoint.split(":");
        return Integer.parseInt(times[0]) * 60 + Integer.parseInt(times[1]);
    }
    public int findMinDifference(List<String> timePoints) {
        List<Integer> minutesList = new ArrayList<>();
        for (String str : timePoints) minutesList.add(transform(str));
        Collections.sort(minutesList, (Integer l, Integer r) -> l - r);
        int minimum = 24 * 60;
        for (int i = 0; i < minutesList.size(); i++) {
            if (i > 0) minimum = Integer.min(minutesList.get(i) - minutesList.get(i - 1), minimum);
            if (i < minutesList.size() - 1) minimum = Integer.min(minutesList.get(i + 1) - minutesList.get(i), minimum);
            int target = 24 * 60 - minutesList.get(i);
            if (i == 0) minimum = Integer.min(minimum, minutesList.get(1) + target);
            else minimum = Integer.min(minimum, minutesList.get(0) + target);
        }
        return minimum;
    }
    
}

File: leetcode/539_Minimum_Time_Difference/539_Minimum_Time_Difference.cpp
class Solution {
public:
    int transform(string& timePoint) {
        return stoi(timePoint.substr(timePoint.find(':') + 1)) + 60 * stoi(timePoint.substr(0, timePoint.find(':')));
    }
    int findMinDifference(vector<string>& timePoints) {
        int timePointsBucket[1440];
        memset(timePointsBucket, 0, 1440 * sizeof(*timePointsBucket));
        for (string str : timePoints) {
            timePointsBucket[transform(str)]++;
        }
        vector<int> timeRes;
        for (int i = 0; i < 1440; i++) {
            if (timePointsBucket[i] > 1) return 0;
            if (timePointsBucket[i] > 0) timeRes.push_back(i);
        }
        int res = 1440;
        for (int i = 0; i < timeRes.size(); i++) {
            if (i > 0) res = min(res, timeRes[i] - timeRes[i - 1]);
            if (i < timeRes.size() - 1) res = min(res, timeRes[i + 1] - timeRes[i]);
            if (i == 0) res = min(res, 1440 - timeRes[i] + timeRes[1]);
            else res = min(res, 1440 - timeRes[i] + timeRes[0]);
        }
        return res;
    }
};
